### PR TITLE
Extract links and type-related methods to their own module.

### PR DESCRIPTION
**Pull request by beauby**

### CODE REVIEW

File: lib/active_model/serializer.rb [hunk 1]
```diff
@@ -17,32 +19,10 @@ class Serializer
     include Associations
     include Attributes
     include Caching
+    include Links
+    include Type
```
Comment (bf4) [inline lib/active_model/serializer.rb]: this pattern will likely bite us if we continue it indefinitely, but I like how it's helping us understand the code better right now

File: lib/active_model/serializer.rb [hunk 2]
```diff
@@ -148,12 +128,6 @@ def read_attribute_for_serialization(attr)
       end
     end
 
-    # @api private
-    # Used by JsonApi adapter to build resource links.
```
Comment (bf4) [inline lib/active_model/serializer.rb]: comment and scope got lost

File: lib/active_model/serializer.rb (hunk 1 above)
Comment (beauby) [inline lib/active_model/serializer.rb]: Well it's quite a common pattern. In any case, it does not make sense to have a 300+ lines `serializer.rb` file.

File: lib/active_model/serializer.rb [hunk 3]
```diff
@@ -17,32 +19,10 @@ class Serializer
...
     require 'active_model/serializer/adapter'
 
-    with_options instance_writer: false, instance_reader: false do |serializer|
-      serializer.class_attribute :_type, instance_reader: true
-      serializer.class_attribute :_links # @api private : links definitions, @see Serializer#link
-      self._links ||= {}
-    end
-
-    # Serializers inherit _attribute_mappings, _reflections, and _links.
-    # Generates a unique digest for each serializer at load.
```
Comment (bf4) [inline lib/active_model/serializer.rb]: I wonder if there's a way to document what's inherited without looking for all the hooks. maybe register our own? dunno

File: lib/active_model/serializer.rb (hunk 2 above)
Comment (beauby) [inline lib/active_model/serializer.rb]: This method actually got replaced by an instance reader on `_links`.

Comment (beauby) [inline lib/active_model/serializer.rb]: https://github.com/rails-api/active_model_serializers/pull/1425/files#diff-e0ecc03de43b196cb38915afa26cd602R8

File: lib/active_model/serializer/links.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
+module ActiveModel
+  class Serializer
+    module Links
+      extend ActiveSupport::Concern
+
+      included do
+        with_options instance_writer: false, instance_reader: true do |serializer|
+          serializer.class_attribute :_links # @api private
```
Comment (bf4) [inline lib/active_model/serializer/links.rb]: So, just added as an instance reader, ok

File: lib/active_model/serializer/links.rb [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+      module ClassMethods
+        def inherited(base)
+          super
+          base._links = _links.dup
+        end
+
+        # Define a link on a serializer.
+        # @example
+        #   link :self { "/posts/#{object.id}" }
```
Comment (bf4) [inline lib/active_model/serializer/links.rb]: missing comma in comment. Also, we should use absolute urls, right?

Comment (beauby) [inline lib/active_model/serializer/links.rb]: No missing comma (this is a block, not a hash). As far as absolute urls are concerned, yes :)

File: lib/active_model/serializer.rb (hunk 3 above)
Comment (beauby) [inline lib/active_model/serializer.rb]: In my point of view it makes sense that each feature handles inheritance for its own data, no?

File: lib/active_model/serializer/links.rb (hunk 5 above)
Comment (bf4) [inline lib/active_model/serializer/links.rb]: > No missing comma (this is a block, not a hash). 

:sadtrombone:

File: lib/active_model/serializer.rb (hunk 1 above)
Comment (bf4) [inline lib/active_model/serializer.rb]: we should just have a reference conversation we write about extracting to modules vs. classes to link to when this comes up :)

File: lib/active_model/serializer.rb (hunk 3 above)
Comment (bf4) [inline lib/active_model/serializer.rb]: yeah.. just makes it harder to chase down by inspecting the code... really just thinking about readability

File: lib/active_model/serializer.rb [hunk 6]
```diff
@@ -17,32 +19,10 @@ class Serializer
     include Associations
     include Attributes
     include Caching
+    include Links
```
Comment (bf4) [inline lib/active_model/serializer.rb]: maybe `include Links # Used by JsonApi adapter to build resource links.`

**Pull request merged**